lib/errorlogger.cpp

```
@@ -945,24 +945,40 @@ std::string getClassification(const std::string &guideline, ReportType reportTyp
         return getClassification(checkers::certCInfo, guideline);
     case ReportType::certCpp:
         return getClassification(checkers::certCppInfo, guideline);
-    case ReportType::misraC:
+    case ReportType::misraC2012:
+    case ReportType::misraC2023:
+    case ReportType::misraC2025:
     {
-        auto components = splitString(guideline, '.');
+        const bool isDirective = guideline.find("Dir ") == 0;
+
+        const std::size_t offset = isDirective ? 4 : 0;
+        auto components = splitString(guideline.substr(offset), '.');
         if (components.size() != 2)
             return "";
 
         const int a = std::stoi(components[0]);
         const int b = std::stoi(components[1]);
 
-        const std::vector<checkers::MisraInfo> &info = checkers::misraC2012Rules;
-        const auto it = std::find_if(info.cbegin(), info.cend(), [&](const checkers::MisraInfo &i) {
+        const std::vector<checkers::MisraInfo> *info = nullptr;
+        switch (reportType) {
+        case ReportType::misraC2012:
+            info = isDirective ? &checkers::misraC2012Directives : &checkers::misraC2012Rules;
+            break;
+        case ReportType::misraC2023:
+            info = isDirective ? &checkers::misraC2023Directives : &checkers::misraC2023Rules;
+            break;
+        case ReportType::misraC2025:
+            info = isDirective ? &checkers::misraC2025Directives : &checkers::misraC2025Rules;
+            break;
+        default:
```

if this default is executed there will be null pointer dereference right? Does Cppcheck warn about it? Otherwise it seems like it would be a good idea to create a false negative ticket..

---

The inner switch is guarded by the outer switch:
```c++
    case ReportType::misraC2012:
    case ReportType::misraC2023:
    case ReportType::misraC2025:
```

---

hmm yes but if we would add misraC2027 in the outer switch and forget this one there will be null pointer dereference.

---

Yep, that's the reason I put it there, but that's a good point. Should we throw an internal error there?

---

that `cppcheck::unreachable()` looks like a really good solution to my concern.

That will show us and Cppcheck what your intention is (that it is NOT executed).

And there is actually a possibility for a compiler/tool to warn if "misra c 2027" is added later in the outer switch but forgotten here in the inner switch (but as far as I know Cppcheck will not warn). So with that `cppcheck::unreachable()` I don't feel the comment is required.

---

Ok, I'll remove the comment!